### PR TITLE
Using private key from an OpenPGP capable smartcard

### LINKED ISSUE #21: Add Yubikey Support
Add Yubikey support to create and use private keys on the hardware.

### PR DESCRIPTION
**Pull request by evik42**
Resolves #21

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -28,6 +28,10 @@ pub enum Error {
     NodeSignNotPermitted,
     #[error("Invalid signature")]
     InvalidSignature,
+    #[error("Invalid signature value (cannot deserialize)")]
+    InvalidSignatureValue,
+    #[error("Invalid public key")]
+    InvalidPublicKey,
```
Comment (MrDarthShoe) [inline src/error.rs:34]: I think there should be test added which expects this specific error

**Review by MrDarthShoe — COMMENTED** [review]
I wonder: maybe we should have some "e2e-like" CLI scenarios added for testing?
i.e.: have root cert then sign it and verify if it is good.
Or have root cert, sign with it some leaf or node-descriptor and verify

File: cli/README.md [hunk 2]
```diff
@@ -46,3 +46,30 @@ The output is saved on the input file path with the extension set to `.signed.js
 ### verify
 
 The command allows verification of certificate or node descriptor JSON files. Apart from the document to be verified an optional timestamp in RFC 3339 format (ex: 2020-01-01T13:42:32Z) can be provided or 'now' to refer to the current time. If the timestamp is provided the document will be verified (beside all other verification) if it is valid at the point of time (the timestamp is within the validity period of the document).
+
+## Smartcard Support
+
+To CLI can utilize smartcards that support OpenPGP with ed25519 signature scheme. This capability is enabled with the `smartcard` feature.
```
[the remark this answers is not in the document]
Comment (evik42) [reply cli/README.md:52]: You would need a hardware smartcard to test.
I added smartcard signed certificates to validator tests.

File: src/cryptography.rs [hunk 3]
```diff
@@ -43,6 +45,41 @@ pub struct Key {
     parameters: Option<Value>,
 }
 
+impl Key {
```
[the remark this answers is not in the document]
Comment (evik42) [reply src/cryptography.rs:48]: good catch!

File: cli/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (evik42) [reply cli/README.md:52]: I did test the code manually.

**Comment by evik42** [comment]
> I wonder: maybe we should have some "e2e-like" CLI scenarios added for testing? i.e.: have root cert then sign it and verify if it is good. Or have root cert, sign with it some leaf or node-descriptor and verify

To test this you would need a build server with a smartcard.
I tested the code manually and also added tests to the certificate validator tests that use the new signature schema.

**Review by MrDarthShoe — APPROVED** [review]

**Comment by MrDarthShoe** [comment]
> > I wonder: maybe we should have some "e2e-like" CLI scenarios added for testing? i.e.: have root cert then sign it and verify if it is good. Or have root cert, sign with it some leaf or node-descriptor and verify
> 
> To test this you would need a build server with a smartcard. I tested the code manually and also added tests to the certificate validator tests that use the new signature schema.

I meant to create test-cases for cli not only for smartcard, but for regular signatures also.
But definitely it can be done in other task

**Pull request merged**